File: crates/tauri-store-cli/assets/plugin-rs/commands.rs
use crate::manager::ManagerExt;
use __IMPORT_SOURCE__::{Result, SaveStrategy, StoreId, StoreOptions, StoreState};
use std::path::PathBuf;
use std::time::Duration;
use tauri::async_runtime::spawn_blocking;
use tauri::{AppHandle, Manager, Runtime, WebviewWindow};

#[tauri::command]
pub(crate) async fn allow_save<R>(app: AppHandle<R>, ids: Vec<StoreId>)
where
  R: Runtime,
{
  let __STORE_COLLECTION__ = app.__STORE_COLLECTION__();
  ids
    .iter()
    .for_each(|id| __STORE_COLLECTION__.allow_save(id));
}

#[tauri::command]
pub(crate) async fn allow_sync<R>(app: AppHandle<R>, ids: Vec<StoreId>)
where
  R: Runtime,
{
  let __STORE_COLLECTION__ = app.__STORE_COLLECTION__();
  ids
    .iter()
    .for_each(|id| __STORE_COLLECTION__.allow_sync(id));
}

#[tauri::command]
pub(crate) async fn clear_autosave<R>(app: AppHandle<R>)
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().clear_autosave();
}

#[tauri::command]
pub(crate) async fn deny_save<R>(app: AppHandle<R>, ids: Vec<StoreId>)
where
  R: Runtime,
{
  let __STORE_COLLECTION__ = app.__STORE_COLLECTION__();
  ids
    .iter()
    .for_each(|id| __STORE_COLLECTION__.deny_save(id));
}

#[tauri::command]
pub(crate) async fn deny_sync<R>(app: AppHandle<R>, ids: Vec<StoreId>)
where
  R: Runtime,
{
  let __STORE_COLLECTION__ = app.__STORE_COLLECTION__();
  ids
    .iter()
    .for_each(|id| __STORE_COLLECTION__.deny_sync(id));
}

#[tauri::command]
pub(crate) async fn get_default_save_strategy<R>(app: AppHandle<R>) -> SaveStrategy
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().default_save_strategy()
}

#[tauri::command]
pub(crate) async fn get_store_collection_path<R>(app: AppHandle<R>) -> PathBuf
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().path()
}

#[tauri::command]
pub(crate) async fn get_store_ids<R>(app: AppHandle<R>) -> Vec<StoreId>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().ids()
}

#[tauri::command]
pub(crate) async fn get_store_path<R>(app: AppHandle<R>, id: StoreId) -> Result<PathBuf>
where
  R: Runtime,
{
  app
    .__STORE_COLLECTION__()
    .with_store(id, |store| store.path())
}

#[tauri::command]
pub(crate) async fn get_save_strategy<R>(app: AppHandle<R>, id: StoreId) -> Result<SaveStrategy>
where
  R: Runtime,
{
  app
    .__STORE_COLLECTION__()
    .with_store(id, |store| store.save_strategy())
}

#[tauri::command]
pub(crate) async fn get_store_state<R>(app: AppHandle<R>, id: StoreId) -> Result<StoreState>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().state(id)
}

#[tauri::command]
pub(crate) async fn load<R>(app: AppHandle<R>, id: StoreId) -> Result<StoreState>
where
  R: Runtime,
{
  spawn_blocking(move || {
    app
      .__STORE_COLLECTION__()
      .with_store(id, |store| store.state().clone())
  })
  .await?
}

#[tauri::command]
pub(crate) async fn patch<R>(window: WebviewWindow<R>, id: StoreId, state: StoreState) -> Result<()>
where
  R: Runtime,
{
  let app = window.app_handle();
  let label = window.label().to_owned();
  app
    .__STORE_COLLECTION__()
    .with_store(id, move |store| store.patch_with_source(state, label))?
}

#[tauri::command]
pub(crate) async fn save<R>(app: AppHandle<R>, id: StoreId) -> Result<()>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().save(id)
}

#[tauri::command]
pub(crate) async fn save_all<R>(app: AppHandle<R>) -> Result<()>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().save_all()
}

#[tauri::command]
pub(crate) async fn save_all_now<R>(app: AppHandle<R>) -> Result<()>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().save_all_now()
}

#[tauri::command]
pub(crate) async fn save_now<R>(app: AppHandle<R>, id: StoreId) -> Result<()>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().save_now(id)
}

#[tauri::command]
pub(crate) async fn save_some<R>(app: AppHandle<R>, ids: Vec<StoreId>) -> Result<()>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().save_some(&ids)
}

#[tauri::command]
pub(crate) async fn save_some_now<R>(app: AppHandle<R>, ids: Vec<StoreId>) -> Result<()>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().save_some_now(&ids)
}

#[tauri::command]
pub(crate) async fn set_autosave<R>(app: AppHandle<R>, interval: u64)
where
  R: Runtime,
{
  app
    .__STORE_COLLECTION__()
    .set_autosave(Duration::from_millis(interval));
}

#[tauri::command]
pub(crate) async fn set_store_collection_path<R>(app: AppHandle<R>, path: PathBuf) -> Result<()>
where
  R: Runtime,
{
  spawn_blocking(move || app.__STORE_COLLECTION__().set_path(path)).await?
}

#[tauri::command]
pub(crate) async fn set_save_strategy<R>(
  app: AppHandle<R>,
  id: StoreId,
  strategy: SaveStrategy,
) -> Result<()>
where
  R: Runtime,
{
  app
    .__STORE_COLLECTION__()
    .with_store(id, |store| store.set_save_strategy(strategy))
}

#[tauri::command]
pub(crate) async fn set_store_options<R>(
  window: WebviewWindow<R>,
  id: StoreId,
  options: StoreOptions,
) -> Result<()>
where
  R: Runtime,
{
  let app = window.app_handle();
  let label = window.label().to_owned();
  app
    .__STORE_COLLECTION__()
    .with_store(id, move |store| {
      store.set_options_with_source(options, label)
    })?
}

#[tauri::command]
pub(crate) async fn unload<R>(app: AppHandle<R>, id: StoreId) -> Result<()>
where
  R: Runtime,
{
  app.__STORE_COLLECTION__().unload_store(&id)
}

File: crates/tauri-store-cli/assets/plugin-ts/commands.ts
// @ts-nocheck
import { commands } from '@tauri-store/shared';

/**
 * Removes one or more stores from the save denylist.
 */
export const allowSave = commands.allowSave('__PLUGIN_TITLE__');

/**
 * Removes one or more stores from the sync denylist.
 */
export const allowSync = commands.allowSync('__PLUGIN_TITLE__');

/**
 * Stops the autosave.
 */
export const clearAutosave = commands.clearAutosave('__PLUGIN_TITLE__');

/**
 * Adds one or more stores to the save denylist.
 */
export const denySave = commands.denySave('__PLUGIN_TITLE__');

/**
 * Adds one or more stores to the sync denylist.
 */
export const denySync = commands.denySync('__PLUGIN_TITLE__');

/**
 * Gets the default save strategy for the stores.
 * It can be overridden on a per-store basis.
 */
export const getDefaultSaveStrategy = commands.getDefaultSaveStrategy('__PLUGIN_TITLE__');

/**
 * Directory where the stores are saved.
 */
export const getStoreCollectionPath = commands.getStoreCollectionPath('__PLUGIN_TITLE__');

/**
 * Gets the save strategy used by a store.
 **/
export const getSaveStrategy = commands.getSaveStrategy('__PLUGIN_TITLE__');

/**
 * Lists all the store ids.
 */
export const getStoreIds = commands.getStoreIds('__PLUGIN_TITLE__');

/**
 * Path where the store is saved.
 **/
export const getStorePath = commands.getStorePath('__PLUGIN_TITLE__');

/**
 * Gets the state of a store.
 */
export const getStoreState = commands.getStoreState('__PLUGIN_TITLE__');

/**
 * Saves a store to the disk.
 *
 * @example
 * ```ts
 * import { save } from '__PACKAGE_NAME__';
 *
 * // Save a single store.
 * await save('my-store');
 *
 * // Save some stores.
 * await save('my-store', 'my-store-2');
 * ```
 **/
export const save = commands.save('__PLUGIN_TITLE__');

/**
 * Saves all the stores to the disk.
 */
export const saveAll = commands.saveAll('__PLUGIN_TITLE__');

/**
 * Saves all the stores to the disk immediately, ignoring the save strategy.
 */
export const saveAllNow = commands.saveAllNow('__PLUGIN_TITLE__');

/**
 * Saves a store to the disk immediately, ignoring the save strategy.
 *
 * @example
 * ```ts
 * import { saveNow } from '__PACKAGE_NAME__';
 *
 * // Save a single store.
 * await saveNow('my-store');
 *
 * // Save some stores.
 * await saveNow('my-store', 'my-store-2');
 * ```
 */
export const saveNow = commands.saveNow('__PLUGIN_TITLE__');

/**
 * Saves the stores periodically.
 * @param interval The interval in milliseconds.
 *
 * @example
 * ```ts
 * import { setAutosave } from '__PACKAGE_NAME__';
 *
 * // Save all the stores every 5 minutes.
 * await setAutosave(5 * 60 * 1000);
 * ```
 */
export const setAutosave = commands.setAutosave('__PLUGIN_TITLE__');

/**
 * Sets the save strategy for a store.
 * Calling this will abort any pending save operation.
 *
 * @example
 * ```ts
 * import { setSaveStrategy } from '__PACKAGE_NAME__';
 *
 * await setSaveStrategy('my-store', 'debounce', 1000);
 * ```
 */
export const setSaveStrategy = commands.setSaveStrategy('__PLUGIN_TITLE__');

/**
 * Sets the store options.
 *
 * @example
 * ```ts
 * import { setStoreOptions } from '__PACKAGE_NAME__';
 *
 * await setStoreOptions('my-store', {
 *  saveOnChange: true,
 *  saveStrategy: 'debounce',
 *  saveInterval: 1000
 * });
 * ```
 */
export const setStoreOptions = commands.setStoreOptions('__PLUGIN_TITLE__');

/////////////////////////
// INTERNAL COMMANDS
/////////////////////////

/** @internal */
export const destroy = commands.destroy('__PLUGIN_TITLE__');

/** @internal */
export const load = commands.load('__PLUGIN_TITLE__');

/** @internal */
export const patch = commands.patch('__PLUGIN_TITLE__');

/** @internal */
export const unload = commands.unload('__PLUGIN_TITLE__');
